fem/tmop_pa.hpp

```diff
@@ -0,0 +1,153 @@
+         typename Kernel_t = typename K::Kernel_t>
+void Emplace(std::unordered_map<Key_t, Kernel_t> &map)
+{
+   constexpr Key_t key = K::template GetKey<N>();
+   constexpr Kernel_t ker = K::template GetKer<key>();
+   map.emplace(key, ker);
+}
+
+/// Instances
+template<class K, typename T, T... idx>
+struct Instances
+{
+   static void Fill(std::unordered_map<typename K::Key_t,
+                    typename K::Kernel_t> &map)
+   {
+      using unused = int[];
+      (void) unused {0, (kernels::Emplace<K,idx>(map), 0)... };
+   }
+};
+
+/// Cat instances
+template<class K, typename Offset, typename LHS, typename RHS> struct Cat;
+template<class K, typename T, T Offset, T... LHS, T... RHS>
+struct Cat<K, std::integral_constant<T, Offset>,
+          Instances<K, T, LHS...>, Instances<K, T, RHS...> >
+{ using Type = Instances<K, T, LHS..., (Offset + RHS)...>; };
+
+/// Sequence
+template<class K, typename T, typename N>
+struct Sequence
+{
+   using LHS = std::integral_constant<T, N::value/2>;
+   using RHS = std::integral_constant<T, N::value-LHS::value>;
+   using Type = typename kernels::Cat<K,LHS,
+         typename Sequence<K,T,LHS>::Type,
+         typename Sequence<K,T,RHS>::Type>::Type;
+};
+
+/// Sequence: empty
+template<class K, typename T>
+struct Sequence<K,T,std::integral_constant<T,0> >
+{ using Type = Instances<K,T>; };
+
+/// Sequence: terminal cases
+template<class K, typename T>
+struct Sequence<K,T,std::integral_constant<T,1> >
+{ using Type = Instances<K,T,0>; };
+
+/// Kernel Make_sequence
+template<class Instance, typename T = typename Instance::Key_t>
+using MakeSequence =
+   typename Sequence<Instance,T,std::integral_constant<T,Instance::N> >::Type;
+
+
+/// Kernel Instantiator class which creates an unordered_map of the Keys/Kernels
+/// declared within the Instance.
+template<class Instance,
+         typename Key_t = typename Instance::Key_t,
+         typename Return_t = typename Instance::Return_t,
+         typename Kernel_t = typename Instance::Kernel_t>
+class Instantiator
+{
+private:
+   using map_t = std::unordered_map<Key_t, Kernel_t>;
+   map_t map;
+
+public:
+   // Fill all the map with the Keys/Kernels declared in the Instance
+   Instantiator() { kernels::MakeSequence<Instance>().Fill(map); }
+
+   bool Find(const Key_t id) { return map.find(id) != map.end(); }
+
+   Kernel_t At(const Key_t id) { return map.at(id); }
+};
+
+// MFEM_REGISTER_TMOP_KERNELS macro:
+//  - the first argument (return_t) is the return type of the kernel
+//  - the second argument (kernel) is the name of the kernel
+//  - the arguments of the kernel (...) captured as __VA_ARGS__
+//
+// This call will output the followings:
+//  1. forward declaration of the kernel
+//  2. kernel pointer declaration
+//  3. struct K##name##_T definition which holds the keys/kernels instance
+//  4. Instantiator definition of the current kernel
+//  5. the kernel signature by re-using all the arguments
```

Comment:
Maybe that adding an example of the result of this macro would help.

Reply:
`done`